src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,311 @@
+        let display_object = display::object::Instance::new(&logger);
+        let size           = Cell::new(Vector2::zero());
+        let gap_width      = Cell::new(SEGMENT_GAP_WIDTH);
+
+        let mut ports      = Vec::default();
+        ports.resize_with(number_of_ports as usize,|| component::ShapeView::new(&logger,&scene));
+        let ports          = RefCell::new(ports);
+
+        OutPutPortsData {display_object,logger,size,ports,gap_width}.init()
+    }
+    fn init(self) -> Self {
+        self.update_shape_layout_based_on_size();
+        self
+    }
+
+    fn update_shape_layout_based_on_size(&self) {
+        let port_num    = self.ports.borrow().len() as f32;
+        let width       = self.size.get().x;
+        let height      = self.size.get().y;
+        let port_width  = (width - NODE_SHAPE_PADDING) / port_num;
+        let port_size   = Vector2::new(port_width, height);
+        let gap_width   = self.gap_width.get();
+
+        // Align shapes along width.
+        let x_start = -width / 2.0 + NODE_SHAPE_PADDING;
+        let x_delta = port_width;
+        for (index, view) in self.ports.borrow().iter().enumerate(){
+            view.display_object().set_parent(&self.display_object);
+
+            let pos_x = x_start + x_delta * index as f32;
+            let pos_y = 0.0;
+            let pos   = Vector2::new(pos_x,pos_y);
+            view.set_position_xy(pos);
+
+            let shape = &view.shape;
+            shape.sprite.size().set(port_size);
+            shape.shape_width.set(width);
+            shape.padding.set(gap_width);
+            shape.offset_x.set(x_delta * index as f32);
+        }
+    }
+
+    fn set_size(&self, size:Vector2<f32>) {
+        self.size.set(size);
+        self.update_shape_layout_based_on_size();
+    }
+}
+
+
+
+// ===================
+// === OutPutPorts ===
+// ===================
+
+/// Implements the segmented output port area. Provides shapes that can be attached to a `Node` to
+/// add an interactive area with output ports.
+///
+/// The `OutputPorts` facilitate the falling behaviour:
+///  * when one of the output ports is hovered, after a set time, all ports are show and the hovered
+///    port is highlighted.
+///  * when a different port is hovered, it is highlighted immediately.
+///  * when none of the ports is hovered all of the `OutputPorts` disappear. Note: there is a very
+///    small delay for disappearing to allow for smooth switching between ports.
+///
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    /// The FRP api of the `OutPutPorts`.
+    pub frp            : Frp,
+        network        : frp::Network,
+        data           : Rc<OutPutPortsData>,
+}
+
+impl OutputPorts {
+    /// Constructor.
+    pub fn new(scene:&Scene, number_of_ports:u32) -> Self {
+
+        let network        = default();
+        let frp            = Frp::new(&network);
+        let data           = OutPutPortsData::new(scene.clone_ref(),number_of_ports);
```

Comment:
If `OutPutPortsData::new` requires the user of its API to clone the argument which implements `CloneRef`, then it is badly designed. Instead, `OutPutPortsData::new` should consume reference and clone it for itself.

Reply:
I think in general that is the preferred way to do this in Rust. The caller might have a spare owned version that they don't need and can themselve decide how and when to create something new.